examples/vision/probing_vits.py

```diff
@@ -0,0 +1,560 @@
+    attentions = attention_scores[0, :, 0, num_tokens:].reshape(num_heads, -1)
+
+    # Reshape the attention scores to resemble mini patches.
+    attentions = attentions.reshape(num_heads, w_featmap, h_featmap)
+    attentions = attentions.transpose((1, 2, 0))
+
+    # Resize the attention patches to 224x224 (224: 14x16).
+    attentions = tf.image.resize(
+        attentions, size=(h_featmap * PATCH_SIZE, w_featmap * PATCH_SIZE)
+    )
+    return attentions
+
+
+"""
+We can use the same image we used for inference with DINO and the `attention_score_dict`
+we extracted from the results.
+"""
+
+# De-normalize the image for visual clarity.
+in1k_mean = tf.constant([0.485 * 255, 0.456 * 255, 0.406 * 255])
+in1k_std = tf.constant([0.229 * 255, 0.224 * 255, 0.225 * 255])
+preprocessed_img_orig = (preprocessed_image * in1k_std) + in1k_mean
+preprocessed_img_orig = preprocessed_img_orig / 255.0
+preprocessed_img_orig = tf.clip_by_value(preprocessed_img_orig, 0.0, 1.0).numpy()
+
+# Generate the attention heatmaps.
+attentions = attention_heatmap(attention_score_dict, preprocessed_img_orig)
+
+# Plot the maps.
+fig, axes = plt.subplots(nrows=3, ncols=4, figsize=(13, 13))
+img_count = 0
+
+for i in range(3):
+    for j in range(4):
+        if img_count < len(attentions):
+            axes[i, j].imshow(preprocessed_img_orig[0])
+            axes[i, j].imshow(attentions[..., img_count], cmap="inferno", alpha=0.6)
+            axes[i, j].title.set_text(f"Attention head: {img_count}")
+            axes[i, j].axis("off")
+            img_count += 1
+
+"""
+## Notes
+
+* DINO extended the attention heatmap generation process to videos. We also
+[applied](https://github.com/sayakpaul/probing-vits/blob/main/notebooks/dino-attention-map
+s-video.ipynb) our DINO implementation on a series of videos and obtained similar
+results. Here's one such video of attention heatmaps:
+
+    ![dino](https://i.imgur.com/kAShjbs.gif)
+
+* [Raghu et al.](https://arxiv.org/abs/2108.08810) use an array of techniques to
+investigate the representations learned by ViTs and make comparisons with that of
+ResNets. We strongly recommend reading their work.
+
+* To author this example, we developed
+[this repository](https://github.com/sayakpaul/probing-vits) to guide our readers so that they
+can easily reproduce the experiments and extend them.
+
+* Another repository that you may find interesting in this regard is
+[`vit-explain`](https://github.com/jacobgil/vit-explain).
+
+* One can also plot the attention rollout and attention heat maps with
+custom images using our Hugging Face spaces.
+
+| Attention Heat Maps | Attention Rollout |
+| :--: | :--: |
+| [![Generic badge](https://img.shields.io/badge/🤗%20Spaces-Attention%20Heat%20Maps-black.svg)](https://huggingface.co/spaces/probing-vits/attention-heat-maps) |
+[![Generic badge](https://img.shields.io/badge/🤗%20Spaces-Attention%20Rollout-black.svg)](https://huggingface.co/spaces/probing-vits/attention-rollout) |
```

Comment:
They even have an emoji in the URL!

Reply:
It was to generate the text on the graphic badge.